meteor/client/lib/forms/schemaFormWithOverrides.tsx

```diff
@@ -0,0 +1,279 @@
+	overrideHelper: OverrideOpHelper
+}
+
+interface FormComponentProps {
+	schema: JSONSchema
+	translationNamespaces: string[]
+	commonAttrs: {
+		label: string
+		hint: string | undefined
+
+		item: WrappedOverridableItemNormal<any>
+		overrideHelper: OverrideOpHelper
+		itemKey: string
+		opPrefix: string
+	}
+}
+
+function useChildPropsForFormComponent(props: SchemaFormWithOverridesProps) {
+	return useMemo(() => {
+		const title = props.schema[SchemaFormUIField.Title] || props.attr
+		const description = props.schema[SchemaFormUIField.Description]
+
+		return {
+			schema: props.schema,
+			translationNamespaces: props.translationNamespaces,
+			commonAttrs: {
+				label: translateStringIfHasNamespaces(title, props.translationNamespaces),
+				hint: description ? translateStringIfHasNamespaces(description, props.translationNamespaces) : undefined,
+				item: props.item,
+				itemKey: props.attr,
+				opPrefix: props.item.id,
+				overrideHelper: props.overrideHelper,
+			},
+		}
+	}, [props.schema, props.translationNamespaces, props.attr, props.item, props.overrideHelper])
+}
+
+export function SchemaFormWithOverrides(props: SchemaFormWithOverridesProps) {
+	const { t } = useTranslation()
+
+	const childProps = useChildPropsForFormComponent(props)
+
+	switch (props.schema.type) {
+		case TypeName.Array:
+			return <ArrayFormWithOverrides {...props} />
+		case TypeName.Object:
+			if (props.schema[SchemaFormUIField.DisplayType] === 'json') {
+				return <JsonFormWithOverrides {...childProps} />
+			} else {
+				return <ObjectFormWithOverrides {...props} />
+			}
+		case TypeName.Integer:
+			if (props.schema.enum) {
+				return <EnumFormWithOverrides {...childProps} />
+			} else {
+				return <IntegerFormWithOverrides {...childProps} />
+			}
+		case TypeName.Number:
+			return <NumberFormWithOverrides {...childProps} />
+		case TypeName.Boolean:
+			return <BooleanFormWithOverrides {...childProps} />
+		case TypeName.String:
+			if (props.schema.enum) {
+				return <EnumFormWithOverrides {...childProps} />
+			} else {
+				return <StringFormWithOverrides {...childProps} />
+			}
+		default:
+			return <>{t('Unsupported field type "{{ type }}"', { type: props.schema.type })}</>
+	}
+}
+
+const ArrayFormWithOverrides = (props: SchemaFormWithOverridesProps) => {
+	const { t } = useTranslation()
+
+	const childProps = useChildPropsForFormComponent(props)
+
+	switch (props.schema.items?.type) {
+		case TypeName.String:
+			return <StringArrayFormWithOverrides {...childProps} />
+		case TypeName.Object:
+			if (props.allowTables) {
+				return <SchemaFormTable {...props} />
+			} else {
+				return <>{t('Tables are not supported here')}</>
+			}
+		default:
+			return <>{t('Unsupported array type "{{ type }}"', { type: props.schema.items?.type })}</>
+	}
+}
+
+const ObjectFormWithOverrides = (props: SchemaFormWithOverridesProps) => {
+	return (
+		<>
+			{' '}
+			{Object.entries(props.schema.properties || {}).map(([index, schema]) => {
```

Comment:
Shouldn't this also take `additionalProperties` into consideration? I'm looking into how this affects Input Gateway, and have one instance where for one of the SubDevice types, I want to have a configuration of an object type, with user-configurable keys that then contains an object definition for that key that varies on a `type` discriminator property. In that case I used a `ConfigManifestEntryType.TABLE` with a `defaultType`, `typeField` and some `config` type definitions. I don't think this is possible here?

Reply:
I dont think `additionalProperties` is something the UI should be looking at. https://json-schema.org/understanding-json-schema/reference/object.html#additional-properties
The best we would be able to do if using it is allow for adding custom fields, using `additionalProperties` and `patternProperties` to limit the names/types of those. But they wont have a description or anything, so that wont be a good replication of what you were doing in input gateway.

I forgot to check input-gateway, so didn't realise that complex use of nested tables was being used. It could either be reimplemented using [`oneOf`](https://json-schema.org/understanding-json-schema/reference/combining.html?highlight=oneof#oneof), or that table could be split up into multiple tables to not need the complex types

Reply:
Yeah, I guess we can live with a multiple tables, I don't expect us to configure the MIDI integration all that often. Although I think we should leave some _TODO_ here, since this is quite a significant limitation, I think.